docs(api/dart): add support for generation and display

chalin opened the pull request:
Fixes #1880, which has a detailed task breakdown.
Supersedes #1593.

Note about the API subsite design: I had to resort to using Jade extends/includes; hence disabling Harp _layout/partials.

cc: @naomiblack @kwalrath @ferhatb @keertip

chalin commented:
@wardbell @Foxandxss : ready for review.

chalin commented:
I forgot to mention that to build the docs do: an install, patch terraforma, and then check-deploy:
```bash
./scripts/install.sh
./scripts/patch.sh
gulp check-deploy --dgeni-log=warn  # drop the flag if you want `info` level messages as usual
```

kwalrath commented:
Taking a look... First impression: nice!

The second thing I noticed is the lack of a language dropdown at /docs/dart/latest/api/ (or any page under it). Would it be possible to add it? (And add some reasonable behavior in the likely case that there's no equivalent page at the dart -> ts/js version of the URL.)

If a dropdown isn't practical, we should put "Dart" in that space, so it's easier to tell that you're in the Dart version of the API docs.

chalin commented:
I chose to leave out the dropdown because IMHO maintaining a complete dart <-> ts/js mapping might be tricky and not really worth the effort (we can open a separate issue as a "feature request" so that we can track discussion about it). Leaving the dropdown there with only Dart is a good idea. I'll look into it.

chalin commented:
@filipesilva @Foxandxss @wardbell : I just rebased. Note the following change:
- Renamed `'harp-compile'` to `'check'` (as in `check-deploy` without the `deploy` :).
- Created new `'harp-compile'` that _only_ does a harp compile; i.e., doesn't depend on `'build-docs'` (where as `'check'` does).

This makes it easier to selectively rebuild only parts of the docs and then quickly get harp to start compiling.

filipesilva commented:
@chalin that change lgtm

chalin commented:
@kwalrath : I've created #1895 for the API page language-cross-referencing issue. I'd like to suggest that nothing be done in the scope of this PR. We can explore the range of alternative solutions via #1895 and a follow up PR. Thoughts?

kwalrath commented:
LGTM.

kwalrath commented:
@filipesilva @Foxandxss @wardbell @naomiblack: Unless you object, I plan to merge this after confirming that I can, indeed, generate the docs. 2 pm Pacific time is my goal.

wardbell commented:
Please **no gulp task called _check_**  Check _what_? I don't approve.  

I do understand the impulse. It comes down to naming. I strongly prefer one of these alternatives (in my preferred order)

1. an option on `harp-compile` such as `harp-compile --no-build`
1. new  `build-compile` task that is `check-deploy` without the deploy.

wardbell commented on gulpfile.js:
```diff
@@ -596,10 +643,35 @@ gulp.task('git-changed-examples', ['_shred-devguide-examples'], function(){
+  return execPromise('npm run live-server ./www');
+});
+
+gulp.task('check', ['build-docs'], function() {
```
This could be `build-compile`

wardbell commented:
Please also update the READM.md to explain these gulp tasks-and-options.

Otherwise, LGTM

chalin commented:
@wardbell : named changed to `build-compile`.

More install / travis / gulpfile changes are on their way (to support Dart API docs building). I'll update the README then.

wardbell commented:
LGTM. @kwalrath is free to merge :+1:

chalin commented:
Design note (cc @petebacondarwin): while originally I started writing a dgeni package for Dart API processing, I ran into problems -- due to my lack of mastery of the dgeni framework --- and so, in the hopes of making some progress, I switched to building a stand alone transformer from the dgeni package parts. Maybe this can be revised later, or, I might just write a revised API page generator in Dart :).
